### PR TITLE
feat(x/ecocredit): `Retire` ORM

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

adds retire method with ORM

ref: #696 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/core/retire.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import (
+	"context"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ecocreditv1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1beta1"
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+)
+
+func (k Keeper) Retire(ctx context.Context, req *v1beta1.MsgRetire) (*v1beta1.MsgRetireResponse, error) {
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire.go:12]: godoc :pray:

File: x/ecocredit/server/core/retire_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	t.Parallel()
+	s := setupBase(t)
+	s.setupClassProjectBatch(t)
+
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(_ interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = types.NewCoins(types.NewInt64Coin("foo", 20))
+		p.CreditTypes = []*ecocredit.CreditType{{Name: "carbon", Abbreviation: "C", Unit: "tonne", Precision: 6}}
+	}).AnyTimes()
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire_test.go:24]:
```suggestion
	}).Times(0)
```

File: x/ecocredit/server/core/retire_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+func TestRetire_Valid(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	s.setupClassProjectBatch(t)
+
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(_ interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = types.NewCoins(types.NewInt64Coin("foo", 20))
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire_test.go:22]: Do we need to include these?

File: x/ecocredit/server/core/retire_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	bal, err := s.stateStore.BatchBalanceStore().Get(s.ctx, s.addr, 1)
+	assert.NilError(t, err)
+	assert.Equal(t, bal.Tradable, "0.5")
+	assert.Equal(t, bal.Retired, "20.5")
+
+	sup, err := s.stateStore.BatchSupplyStore().Get(s.ctx, 1)
+	assert.NilError(t, err)
+	assert.Equal(t, sup.TradableAmount, "0.5")
+	assert.Equal(t, sup.RetiredAmount, "20.5")
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire_test.go:43]: Would be nice to have comments with original values and how they get updated as you did on your send pr.

File: x/ecocredit/server/core/retire_test.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+			{BatchDenom: "A00-00000000-00000000-01", Amount: "10.35"},
+		},
+		Location: "US-NY",
+	})
+	assert.ErrorContains(t, err, ormerrors.NotFound.Error())
+
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(_ interface{}, p *ecocredit.Params) {
+		p.AllowlistEnabled = false
+		p.CreditClassFee = types.NewCoins(types.NewInt64Coin("foo", 20))
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire_test.go:64]: Are these necessary to include?

File: x/ecocredit/server/core/retire_test.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		p.CreditTypes = []*ecocredit.CreditType{{Name: "carbon", Abbreviation: "C", Unit: "tonne", Precision: 6}}
+	}).AnyTimes()
```
Comment (ryanchristo) [inline x/ecocredit/server/core/retire_test.go:66]:
```suggestion
	}).Times(0)
```

**Review by ryanchristo — APPROVED** [review]
Looks good. A few minor suggestions. Pre-approving.

**Review by amaurym — APPROVED** [review]
LGTM.

One improvement we could do is group the credits by batch in the loop, and do store reads/writes by batch, to save some gas. But okay to leave it as is for now.

**Pull request merged**